azurelinuxagent/ga/cpucontroller.py

```diff
@@ -0,0 +1,269 @@
+
+    def get_cpu_throttled_time(self, read_previous_throttled_time=True):
+        # Throttled time is reported in nanoseconds in v1
+        if not read_previous_throttled_time:
+            return float(self._get_cpu_stat_counter(counter_name='throttled_time') / 1E9)
+
+        if not self._cpu_usage_initialized():
+            raise CGroupsException("initialize_cpu_usage() must be invoked before the first call to get_cpu_throttled_time()")
+
+        self._previous_throttled_time = self._current_throttled_time
+        self._current_throttled_time = self._get_cpu_stat_counter(counter_name='throttled_time')
+
+        return round(float(self._current_throttled_time - self._previous_throttled_time) / 1E9, 3)
+
+
+class CpuControllerV2(_CpuController):
+    def initialize_cpu_usage(self):
+        if self._cpu_usage_initialized():
+            raise CGroupsException("initialize_cpu_usage() should be invoked only once")
+        self._current_cgroup_cpu = self._get_cpu_time(allow_no_such_file_or_directory_error=True)
+        self._current_system_cpu = self._osutil.get_system_uptime()
+        self._current_throttled_time = self._get_cpu_stat_counter(counter_name='throttled_usec')
+
+    def _get_cpu_time(self, allow_no_such_file_or_directory_error=False):
+        """
+        Returns the CPU time (user and system) consumed by this cgroup in seconds.
+
+        If allow_no_such_file_or_directory_error is set to True and cpu.stat does not exist the function
+        returns 0; this is useful when the function can be called before the cgroup has been created.
+        """
+        try:
+            cpu_stat = self._get_file_contents('cpu.stat')
+        except Exception as e:
+            if not isinstance(e, (IOError, OSError)) or e.errno != errno.ENOENT:  # pylint: disable=E1101
+                raise CGroupsException("Failed to read cpu.stat: {0}".format(ustr(e)))
+            if not allow_no_such_file_or_directory_error:
+                raise e
+            cpu_stat = None
+
+        cpu_time = 0
+
+        if cpu_stat is not None:
+            #
+            # Sample file:
+            #     # cat /sys/fs/cgroup/azure.slice/azure-walinuxagent.slice/azure-walinuxagent-logcollector.slice/collect-logs.scope/cpu.stat
+            #     usage_usec 1990707
+            #     user_usec 1939858
+            #     system_usec 50848
+            #     core_sched.force_idle_usec 0
+            #     nr_periods 397
+            #     nr_throttled 397
+            #     throttled_usec 37994949
+            #     nr_bursts 0
+            #     burst_usec 0
+            #
+            match = re_v2_usage_time.match(cpu_stat)
+            if not match:
+                raise CGroupsException("The contents of {0} are invalid: {1}".format(self._get_cgroup_file('cpu.stat'), cpu_stat))
+            cpu_time = int(match.groups()[0]) / 1E6
+
+        return cpu_time
+
+    def get_cpu_usage(self):
```

Comment:
get_cpu_usage looks about same for both v1 and v2, can we move to abstract class?

Reply:
They're slightly different, most notably we don't multiply by # of cores for v2 (line 256)

Reply:
why don't multiply by # no of cores

Reply:
From my testing multiplying by # of cores gave inaccurate results for v2. I think may have to do that we're getting total system usage from /proc/uptime